### PR TITLE
Dev gfdl main candidate 2021 07 08

### PR DESCRIPTION
**Pull request by marshallward**
This release includes several new features and bugfixes to MOM6.

General improvements include updates to the sponges, surface forcing,
diagnostics, and the ice shelf.

Many issues related to FMS2 support have been resolved, particularly related to
handling of restart files.

Note that the `input_filename` paramenter of the SIS2 namelist group was not
being respected in many cases.  It is now being enforced.  Coupled experiments
which were not managing this parameter should confirm that sea ice fields are
being appropriate initialized after restart.

This patch applies stricter memory management to the source code, in
preparation for more aggressive valgrind testing.  This may cause unexpected
behavior in drivers and coupled models which reference MOM fields outside of
MOM functions.

----

Review
------

- :heavy_check_mark: COAPS / FSU: @abozec
- :heavy_check_mark: EMC: @jiandewang
- :heavy_check_mark: ESMG: @kshedstrom
- :heavy_check_mark: NASA-GMAO: @sanAkel
- :heavy_check_mark: NCAR: @alperaltuntas / @gustavo-marques

---

Features
--------
- #1427: Update flux const (@wfcooke)
- #1416: Ice dynamics (@OlgaSergienko)
- #1412: Fix to masking of u,v sponge code. (@kshedstrom)
- #1398: 3D thickness x momentum diagnostics (@hmkhatri)
- #1387: Ice dynamics (@OlgaSergienko)
- #1383: Adapts Phillips case to work in doubly periodic mode (@adcroft)
- #1372: Update of Eady growth rate used in MEKE and Visbeck (@adcroft)

Bugfixes
--------
- #1432: update halo and allow sponge_uv to not be on grid (@abozec)
- #1431: New fix reflect tides (@raphaeldussin)
- #1428: Bugfix: Minor changes on clipping topography from file (@herrwang0)
- #1425: Change the declaration of stack variable in apply_flux_adjustments (@MJHarrison-GFDL)
- #1411: +Dimensional rescaling of user OBC test cases (@Hallberg-NOAA)
- #1409: Ice shelf restart fix (@MJHarrison-GFDL)
- #1407: corrections to beta-plane (@raphaeldussin)
- #1406: Trying to fix Kelvin wave boundary condition. (@kshedstrom)
- #1399: Avoids issue with unassociated axes in d-sample code (@adcroft)
- #1380: read_attribute_str returns cleanly when found=F (@mnlevy1981)
- #1364: Fixed downsampling for x:sum y:point z:point diags (@jkrasting)
- #1356: Bugfix in getting surface currents [Additions from GMAO (3)] (@sanAkel)

Framework
---------
- #1426: (*)Correct multi-name handling by save_restart (@Hallberg-NOAA)
- #1405: Call variable_att_exists before get_variable_att (@Hallberg-NOAA)
- #1392: infra, FMS_cap: stdout_if_root() (@marshallward)
- #1389: Use MOM_io interfaces in write_depth_list (@Hallberg-NOAA)
- #1388: Use MOM_io in Surface_Bands_by_data_override (@Hallberg-NOAA)
- #1373: +Make MOM_read_data work when 4-d arrays exist (@Hallberg-NOAA)
- #1371: (*)Write full checksums to restarts (@Hallberg-NOAA)
- #1370: +More consistent treatment of input_filename = 'F' (@Hallberg-NOAA)
- #1369: +Enhanced support for novel axes in MOM_io (@Hallberg-NOAA)
- #1363: +Add infrastructure interfaces needed by SIS2 (@Hallberg-NOAA)
- #1359: +Enable FMS2 interface writes (@Hallberg-NOAA)
- #1353: Only write checksums from root PE for FMS_cap (@Hallberg-NOAA)
- #1352: +Enable FMS2 interface reads (includes merge of PR #1165) (@Hallberg-NOAA)

Testing
-------
- #1436: Update gitlab pipeline to use scripts from MOM6-examples (@adcroft)
- #1396: Autoconf: FMS 2020.x compatibility testing (diag_axis_mod, data_override_mod) (@marshallward)
- #1386: .testing: Create python venv at compile time (@marshallward)
- #1376: Display hash of submodules in gitlab pipeline (@adcroft)

Refactoring
-----------
- #1435: (+)Modified some optional arguments (@Hallberg-NOAA)
- #1414: Standardize dimensional rescaling in tracer stocks (@Hallberg-NOAA)
- #1413: Memory management cleanup and initialization fixes (@marshallward)
- #1408: Remove duplicated solvers between regrid modules (@angus-g)
- #1404: +Move wave module variables into control structure (@Hallberg-NOAA)
- #1403: +Add dimensional rescaling of wave variables (@Hallberg-NOAA)
- #1402: Add variable and file names to an error message (@Hallberg-NOAA)
- #1390: Collocate units and conversion for register_diag_field calls (@Hallberg-NOAA)
- #1385: (*)Correct inconsistent units in 17 diagnostics (@Hallberg-NOAA)
- #1382: +Correct the units of remap tendency diagnostics (@Hallberg-NOAA)
- #1365: Coriolis: Improved coradcalc vectorization (@marshallward)
- #1362: MOM_hor_visc: Variables moved to stack (@marshallward)
- #1360: Remove inappropriate timelevel arguments (@Hallberg-NOAA)

External
--------
- #1381: Main candidate from NCAR (04/21/2021) (@gustavo-marques)

----

Contributors:
  - @Hallberg-NOAA: 22
  - @marshallward: 6
  - @adcroft: 5
  - @raphaeldussin: 2
  - @MJHarrison-GFDL: 2
  - @OlgaSergienko: 2
  - @kshedstrom: 2
  - @abozec: 1
  - @herrwang0: 1
  - @wfcooke: 1
  - @angus-g: 1
  - @hmkhatri: 1
  - @gustavo-marques: 1
  - @mnlevy1981: 1
  - @jkrasting: 1
  - @sanAkel: 1

### CODE REVIEW

**Comment by jiandewang** [comment]
@marshallward : the problem in MOM_wave_interface.F90 I mentioned at
https://github.com/NOAA-GFDL/MOM6/pull/1429#issuecomment-862825907
has not been solved in this PR. Thus I am unable to compile the code from this PR.

**Comment by marshallward** [comment]
@jiandewang Thanks for the reminder.  I believe that moving the wave fields into the CS (thereby making them private) was intentional, so I will need to check with @Hallberg-NOAA about how this should be fixed.

**Comment by kshedstrom** [comment]
I approve this PR. It reproduces my former restart with ice answers.

File: src/user/tidal_bay_initialization.F90 [hunk 1]
```diff
@@ -44,6 +47,10 @@ function register_tidal_bay_OBC(param_file, CS, OBC_Reg)
   endif
   allocate(CS)
 
+  call get_param(param_file, mdl, "TIDAL_BAY_FLOW", CS%tide_flow, &
+                 "Maximum total tidal volume flux.", &
+                 units="m3 s-1", default=3.0d6, scale=US%m_s_to_L_T*US%m_to_L*US%m_to_Z)
```
Comment (alperaltuntas) [inline src/user/tidal_bay_initialization.F90:52]: With the gnu/10.1.0 compiler, we get a compilation error due to dummy arg specification `default=3.0d6`. Changing that to `default=3.0e6` fixes the issue for us.

```
   52 |                  units="m3 s-1", default=3.0d6, scale=US%m_s_to_L_T*US%m_to_L*US%m_to_Z)
      |                                                                                        1
Error: There is no specific subroutine for the generic ‘get_param’ at (1)
```

Comment (marshallward) [reply src/user/tidal_bay_initialization.F90:52]: I wonder if this is a compiler bug, maybe related to our kind-less `real` declarations and `-r8` flag usage.  Do you see this in 10.2?  I see no issues with 10.2 and 11.1.  Just asking out of interest, no need to investigate further.

Either way, easy enough to fix.

File: src/parameterizations/vertical/MOM_ALE_sponge.F90 [hunk 2]
```diff
@@ -1006,7 +1021,8 @@ subroutine apply_ALE_sponge(h, dt, G, GV, US, CS, Time)
 
       call pass_var(sp_val,G%Domain)
       do j=CS%jsc,CS%jec; do I=CS%iscB,CS%iecB
-       sp_val_u(I,j,1:nz_data) = 0.5*(sp_val(i,j,1:nz_data)+sp_val(i+1,j,1:nz_data))
+        sp_val_u(I,j,1:nz_data) = 0.5*(sp_val(i,j,1:nz_data)+sp_val(i+1,j,1:nz_data))
+        mask_u(I,j,1:nz_data) = max(mask_z(i,j,1:nz_data),mask_z(i+1,j,1:nz_data))
```
Comment (abozec) [inline src/parameterizations/vertical/MOM_ALE_sponge.F90:1025]: Taking the average of the sp_val will potentially include a masked valued (huge value if set that way in the input file) and it will not be masked by taking the max of the mask_z. So I would suggest, if we want to keep the max of mask idea to weigh the sp_val by mask_z:
sp_val_u=(sp_val(i)*mask_z(i) + sp_val(i+1)*mask_z(i+1)) / max(mask_z(i)+mask_z(I+1),1e-10)

Comment (MJHarrison-GFDL) [reply src/parameterizations/vertical/MOM_ALE_sponge.F90:1025]: @abozec This PR does not contain #1434 which redefines mask_uv and I believe resolves masking issues.   Specifically, the sponge_uv is turned off at OBC points and adjacent to land in #1434.  Sponge target masked values in the interior would cause a problem,  but that shouldn't happen.

Comment (marshallward) [reply src/parameterizations/vertical/MOM_ALE_sponge.F90:1025]: I assume the `mask_z` update is following the 0-or-1 convention as other MOM masks, so that `mask_u` is water (=1) if either east/west c-point is water.

I want to believe that a similar masking convention would have been applied to `sp_val` in `horiz_interp_and_extrap_tracer`, but I could be wrong here.

If there is some large value in `sp_val` caused by an input mask value, then perhaps this should be addressed in `horiz_interp_and_extrap_tracer`?

Comment (marshallward) [reply src/parameterizations/vertical/MOM_ALE_sponge.F90:1025]: I see from Matt's comment and #1434 that `max` was flipped to `min`, extending the mask.  I also do not see any special handling of input mask values in `horiz_interp_and_extrap_tracer`.

https://github.com/NOAA-GFDL/MOM6/blob/52f906cfc458ee77630f75c9d65da52a87fcfa70/src/framework/MOM_horizontal_regridding.F90#L891-L900

Maybe this is clobbering out any potentially large values causing by a different input mask value?

Comment (abozec) [reply src/parameterizations/vertical/MOM_ALE_sponge.F90:1025]: That piece of code in MOM_horizontal_regridding does not filter all the land values for u and v, I'm not sure why... so a mask value in sponge seems to be the solution. 
I've tried the max and Matt's min method and both work in my example. It now depends on  if users prefer an extended mask (using min) or using the 1  value available (using max) to calculate the u and v sponge values.

File: src/user/tidal_bay_initialization.F90 (hunk 1 above)
Comment (alperaltuntas) [reply src/user/tidal_bay_initialization.F90:52]: My bad. Adding the `-fdefault-double-8` compiler flag fixes the issue. It's interesting that this dummy arg is the only instance that requires the `-fdefault-double-8` flag, but that's fine.

**Comment by Hallberg-NOAA** [comment]
@jiandewang, after some discussions with @breichl, I think that we may have a PR that fixes the issues with compiling the waves code with the NUOPC cap.  Please test out https://github.com/NOAA-GFDL/MOM6/pull/1443.

File: src/user/tidal_bay_initialization.F90 (hunk 1 above)
Comment (Hallberg-NOAA) [reply src/user/tidal_bay_initialization.F90:52]: Personally I do not see any reason why we should not change the dummy argument specification to `default=3.0e6`, as suggested by @alperaltuntas, to correct the problem without requiring any new compiler flags and to match the pattern used elsewhere in the MOM6 code.  This seems to be a case where correcting the problem twice might be in order.

**Review by sanAkel — APPROVED** [review]
This PR reproduces answers that I obtained from `develop` version of GEOSgcm

**Comment by marshallward** [comment]
The branch has been updated with two new PRs:

- #1442 
- #1443

The latter was submitted to address the issues reported by EMC.

**Comment by jiandewang** [comment]
@marshallward FYI: PR#1443 solved compiling issue but we are still having issue to run the model.

**Comment by marshallward** [comment]
This branch has been updated with the following PR:
- #1446 
This should address the new issue reported by EMC.

**Comment by jiandewang** [comment]
> This branch has been updated with the following PR:
> 
> * [initialize CS%WaveNum_Cen before read in this parameter in MOM_wave_interface.F90 #1446](https://github.com/NOAA-GFDL/MOM6/pull/1446)
> 
> This should address the new issue reported by EMC.

UFS works fine now

File: src/parameterizations/vertical/MOM_ALE_sponge.F90 (hunk 2 above)
Comment (marshallward) [reply src/parameterizations/vertical/MOM_ALE_sponge.F90:1025]: @abozec Is this something you need to be fixed before we merge?  Or can it be addressed in the next update to main?

**Comment by marshallward** [comment]
The last round of updates have resolved the issues seen by EMC.

Could I ask the other groups review the updated candidate?

File: src/parameterizations/vertical/MOM_ALE_sponge.F90 (hunk 2 above)
Comment (abozec) [reply src/parameterizations/vertical/MOM_ALE_sponge.F90:1025]: it can be addressed with #1434 of Matt's so I'm good with that.

Comment (marshallward) [reply src/parameterizations/vertical/MOM_ALE_sponge.F90:1025]: Ok great, we will merge 1434 in the next dev/gfdl -> main merge, which is expected to happen in a couple of weeks.

**Comment by sanAkel** [comment]
> The last round of updates have resolved the issues seen by EMC.
> 
> Could I ask the other groups review the updated candidate?

@marshallward I looked up the changes since my last [check](https://github.com/NOAA-GFDL/MOM6/pull/1440#pullrequestreview-703612027), these changes "do not apply" to GEOS, please ✅ me off!

**Comment by kshedstrom** [comment]
I still approve this PR.

**Comment by abozec** [comment]
I approve also.

**Comment by alperaltuntas** [comment]
I approve as well.

**Comment by marshallward** [comment]
Merge is complete, well done everyone!

We will try to prepare another one soon with the OBC mask fixes.
